### PR TITLE
Yiming pip flow

### PR DESCRIPTION
**Pull request by liuyiming0507**

### CODE REVIEW

File: geo/Readme.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+## Geometry and mesh built in Salome
+The label (mark number) of subdomains (facets and cells) can be found in .msh file
+In my code I use the .xml files. But it cannot be run on the server benz. The problem seems to be that the part of the code where reads the .xml file cannot run.
+## Bernoullie_test:
```
Comment (sebglane) [inline geo/Readme.md]: It is **Bernoulli**

File: python/crusher.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+cells = MeshFunction('size_t', mesh , fname + '_physical_region.xml')
+facets = MeshFunction('size_t', mesh , fname + '_facet_region.xml')
+#nn = FacetNormal(mesh)
+#on the server benz, the xml cannot be read, How to deal with this problem?
+
+i, j, k, l = indices(4)
+delta = Identity(2)
+
+scalar = FiniteElement('P', triangle, 1)
+vector = VectorElement('P', triangle, 1)
```
Comment (sebglane) [inline python/crusher.py:57]: Here you need to use second order polynomials as well.

File: python/crusher.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+#dimensionless quantities
+
+mu = mu0 + (mu1 - mu0)*exp(-lambd*(J-J1)/J1)
+d = as_tensor( 1./2.*(v[i].dx(j)+v[j].dx(i)) , [i,j] )
+tau = as_tensor( 2.*mu*d[i,j] , [i,j] )
+#tau = as_tensor( mu*v[i].dx(j) , [i,j] )
+g = Expression(('0.','-g'),degree=1,g=0.907)
+F_1 = v[i].dx(i)*del_p*dv
+F_2 = ( ( (v-v0)[j]/dt -g[j] + (v[i]*v[j]).dx(i) + p.dx(j) )*del_v[j]  + tau[i,j]*del_v[j].dx(i) )*dv 
+F_3 = ( (J-J0)/dt + v[i]*J.dx(i) + a_*p*(J-1) )*del_J*dv
```
Comment (sebglane) [inline python/crusher.py:125]: I don't think that you can discretize the evolution equation for the inertia tensor in this way. It is known in the literature that a transport equation is numerically unstable and requires some special treatment.

File: python/crusher.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+Form = F_1 + F_2 + F_3
+
+Gain = derivative(Form, unkn, dunkn)
+while t < t_end:
+	t += dt
+	n += 1
+	solve(Form == 0, unkn, bcs=bc, J=Gain, solver_parameters={"newton_solver":{"linear_solver": "mumps", "relative_tolerance": 1E-3,\
+			"krylov_solver": krylov_params, "maximum_iterations":30, "error_on_nonconvergence": True} } )
```
Comment (sebglane) [inline python/crusher.py:133]: Moreover, a `relative_tolerance` of 1.0e-3 is not very accurate. If everything works well in your code you should be able to decrease this tolerance to any value you want, for example, to 1.0e-12.

**Review by sebglane — APPROVED** [review]
I think this is a start but it is to early for the crusher problem because we first need to solve the other problem in a reliable way.

Feel free to merge but I think that this programs will be superseded soon.

**Comment by sebglane** [comment]
> Sorry, I cannot understand this well, I'm still a beginner. When I tried to generate the weak form, I multiplied it by the test function and then integrated it by parts to eliminate the second-order partial derivative of the speed field (to eliminate the partial derivative of the `tau[i,j]`), and I got the equations like this form.
> 
> ```python
> tau[i,j ] * del_v[j].dx(i)
> ```
> 
> I don't know what should I do when I generate the weak form, in order to achieve this form:
> 
> ```python
> tau[i,j ] * 1. / 2. * ( del_v[j].dx(i) + del_v[i].dx(j) )
> ```
> 
> I tried both in the program, and the results are the same. What is the main influence of this part?
> 
> I think index notation is difficult to read and easy to cause confusion, but sometimes I think it can show the structure of the equations directly. I had a little self-conflict.

There are two reasons for choosing the second version:

1. For the double contraction of a symmetric tensor, in this case `tau[i, j]`, with any other second rank tensor only the symmetric part of the other tensor matters, because the antisymmetric part cancels. You can work this out with pen and paper by decomposing the tensor into the sum of a symmetric and an antisymmetric tensor.
2. In the FEM, a system matrix is assembled and the corresponding linear system is solved. The system matrix is symmetric if the weak form is symmetric which means that you can interchange trial and test functions. The reason why you may want a system matrix with as much as symmetries as possible is that these matrices have a lot of beneficial properties. These properties can, for example, be exploited when solving the linear system.

**Comment by sebglane** [comment]
> I should better change the line 44, 45:
> 
> ```python
> scalar = FiniteElement('P', triangle, 1)
> vector = VectorElement('P', triangle, 1)
> ```
> 
> to
> 
> ```python
> scalar = FiniteElement('P', triangle, 1)
> vector = VectorElement('P', triangle, 2)
> ```
> 
> Is that right?
> 
> And in the week form, we only have the first-order derivative of the velocity vector, why do we need to use quadratic elements in vector space?

Yes your code is right. The reason why we need to this is the famous inf-sup condition or LBB condition. The explanation and proof of this condiiton is pure math. But when take you a look at the equations the incompressibility condition adds as an additional constraint to the velocity. This means that we require more degrees of freedom for the velocity. A better explanation with example is given in [a video lecture of Wolfgang Bangerth](https://www.math.colostate.edu/~bangerth/videos.676.33.25.html). [These lectures](https://www.math.colostate.edu/~bangerth/videos.html) are in general very helpful.

File: python/crusher.py (hunk 4 above)
Comment (liuyiming0507) [reply python/crusher.py:133]: You are right, this should be changed to a smaller value, I set a larger value to make the iteration faster

File: python/crusher.py (hunk 3 above)
Comment (liuyiming0507) [reply python/crusher.py:125]: Yes, I have no confidence in this part, I just used the equation directly after generating a weak form

File: python/crusher.py (hunk 2 above)
Comment (liuyiming0507) [reply python/crusher.py:57]: After watching that video, I think I kind of understand why we use second order polynomials here. Thank you so much!

**Comment by sebglane** [comment]
What should we do about this PR?

**Comment by idanven** [comment]
> What should we do about this PR?

I think we can move this PR to another new repo, since it is not an objected-oriented  program, which we can still discuss it. 
but it depends on @liuyiming0507 .

**Comment by liuyiming0507** [comment]
It's ok for me

**Comment by liuyiming0507** [comment]
> > What should we do about this PR?
> 
> I think we can move this PR to another new repo, since it is not an objected-oriented program, which we can still discuss it.
> but it depends on @liuyiming0507 .

It's ok for me. Moving this PR to another new repo, so that we can keep the comments

**Comment by sebglane** [comment]
@idanven @liuyiming0507 Please feel free to move the code somewhere else. I don't know how to do that...

**Pull request closed without merging**